src/blenderbim/blenderbim/core/georeference.py

```diff
@@ -0,0 +1,99 @@
+# BlenderBIM Add-on - OpenBIM Blender Add-on
+# Copyright (C) 2022 Dion Moult <dion@thinkmoult.com>
+#
+# This file is part of BlenderBIM Add-on.
+#
+# BlenderBIM Add-on is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# BlenderBIM Add-on is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with BlenderBIM Add-on.  If not, see <http://www.gnu.org/licenses/>.
+
+import bpy
+import blenderbim.bim.helper
+
+from blenderbim.bim.module.georeference.data import GeoreferenceData
+
+from math import radians, degrees, atan, tan, cos, sin
+
+def add_georeferencing(ifc):
+    ifc.run("georeference.add_georeferencing")
+
+def enable_editing_georeferencing(georeference):
+    props = bpy.context.scene.BIMGeoreferenceProperties
+    
+    georeference.clear_projected_crs()
+    
+    blenderbim.bim.helper.import_attributes(
+        "IfcProjectedCRS",
+        props.projected_crs,
+        GeoreferenceData.data["projected_crs"],
+        georeference.import_projected_crs_attributes,
+    )
+       
+    georeference.clear_map_conversion()
+        
+    blenderbim.bim.helper.import_attributes(
```

Comment:
Don't call blenderbim.bim.helper from the core. The core should only ever call functions from the tool, and nothing else. This way the core can remain abstract and testable without launching Blender. This also forces all behaviour to be encapsulated in functions in a tool, which promotes code reuse.

So maybe something like georeference.import_map_conversion()

Then inside import_map_conversion use blenderbim.bim.helper.import_attributes2() which is simpler to use and you can then remove the need for GeoreferenceData.data["map_conversion"]

Reply:
Ok, i understand. Just to be clear, have i to delete (in "core/georeference.py") also "import bpy" to get the props?

Reply:
Yes, core/georeference.py should never import bpy. Think of any core code as pure Python functions that call tool functions. No knowledge of external libraries like bpy, ifcopenshell, etc. The only purpose is to:

 1. Allow code maintainers to read a short poem that describes what might be a complex operation in simple English
 2. Force code writers to encapsulate all concrete code in well named functions, that promotes maintainability and code reuse
 3. Through testing, quickly identify butterfly effects of changing code that impacts other parts of the system.
 4. Decouple operations from the Blender UI so that operators can be embedded in other operators without incurring the speed penalty from the Blender UI.

Reply:
Ok, so i tried to remove imports from core and put same imports into tool.
The problem is that the import from GeoreferenceData creates a circular import exception.
So, to resolve, i saw what has been done in other module (Owner) and i saw that this problem is resolved avoiding the import GeoreferenceData and read ifc directly.
I'll fix in this way, is it correct?

Reply:
Tool should not import GeoreferenceData, tool should access IFC directly. The GeoreferenceData (and any other Data classes) are only meant to be accessed by the UI. More reading: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/blenderbim/bim/module/demo/data.py#L26-L30

Removing the relationships between Operators and Data was one of the goals of the refactor (which you are doing now), because if Operators and Data were related, then that would force you to know how to refresh all the correct portions of the UI for every operation, which is very difficult to know especially if operations can have side effects (e.g. when you delete an object, should an IfcTask in a schedule refresh its quantities? Etc)